### PR TITLE
TBD: create shard inputs only once per query instead of once per shard

### PR DESCRIPTION
**Pull request by seut**
By using shard based collect expressions, inputs can be
created once per query for all shards, thus avoid allocations.

Benchmarks results are not really showing an improvement but profiling observed a general allocations decrease by ~35% (same spec used like for benchmarks).

### Benchmark Results
```
# Results (server side duration in ms)
V1: 3.1.0-994f0a5e9afaac870e3cd7465792160b8e9e83c9
V2: 3.1.0-fb199ecccd39fb0177806cacc448729a312ed84a
```

#### First Run
```
Q: select * from sys.shards order by schema_name, table_name
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        5.684 ±    1.172 |      4.210 |      5.440 |      5.690 |     19.288 |
|   V2    |        5.851 ±    4.392 |      4.810 |      5.494 |      5.778 |    141.133 |
mean:   +   2.90%
median: +   0.99%
Likely NOT significant

Q: select * from sys.shards order by schema_name, table_name
C: 15
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        7.923 ±    2.665 |      4.350 |      7.050 |      9.335 |     19.346 |
|   V2    |        8.640 ±    3.224 |      4.487 |      7.613 |     10.082 |     28.707 |
mean:   +   8.67%
median: +   7.67%
Likely significant

Q: select * from sys.shards
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        4.998 ±    0.772 |      4.089 |      4.839 |      4.991 |     13.593 |
|   V2    |        4.747 ±    0.759 |      3.915 |      4.592 |      4.782 |     17.350 |
mean:   -   5.15%
median: -   5.23%
Likely significant

Q: select * from sys.shards
C: 15
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        7.724 ±    2.808 |      3.690 |      6.816 |      9.397 |     25.862 |
|   V2    |        7.739 ±    2.160 |      4.358 |      7.241 |      8.735 |     17.435 |
mean:   +   0.20%
median: +   6.05%
Likely NOT significant
```

#### Second Run
```
Q: select * from sys.shards order by schema_name, table_name
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        5.797 ±    1.070 |      4.912 |      5.542 |      5.776 |     17.798 |
|   V2    |        5.687 ±    1.239 |      4.552 |      5.438 |      5.837 |     21.114 |
mean:   -   1.91%
median: -   1.89%
Likely NOT significant

Q: select * from sys.shards order by schema_name, table_name
C: 15
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        8.139 ±    2.909 |      4.383 |      7.198 |      9.552 |     25.834 |
|   V2    |        8.664 ±    3.513 |      4.251 |      7.672 |      9.518 |     47.185 |
mean:   +   6.25%
median: +   6.37%
Likely significant

Q: select * from sys.shards
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        4.948 ±    0.855 |      3.758 |      4.861 |      5.043 |     17.631 |
|   V2    |        4.744 ±    0.684 |      3.895 |      4.588 |      4.808 |     10.988 |
mean:   -   4.20%
median: -   5.77%
Likely significant

Q: select * from sys.shards
C: 15
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        7.897 ±    2.866 |      4.246 |      6.946 |      9.557 |     24.702 |
|   V2    |        8.068 ±    2.633 |      4.265 |      7.416 |      9.263 |     34.585 |
mean:   +   2.14%
median: +   6.54%
Likely NOT significant
```

#### Third Run
```
Q: select * from sys.shards order by schema_name, table_name
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        5.923 ±    5.589 |      4.765 |      5.438 |      5.667 |    176.749 |
|   V2    |        5.838 ±    4.467 |      4.781 |      5.478 |      5.735 |    143.655 |
mean:   -   1.45%
median: +   0.74%
Likely NOT significant

Q: select * from sys.shards order by schema_name, table_name
C: 15
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        8.062 ±    2.827 |      4.365 |      7.088 |      9.395 |     22.698 |
|   V2    |        8.861 ±    3.724 |      4.604 |      7.713 |     10.351 |     49.856 |
mean:   +   9.44%
median: +   8.45%
Likely significant

Q: select * from sys.shards
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        4.961 ±    0.812 |      4.100 |      4.782 |      4.976 |     17.859 |
|   V2    |        4.793 ±    0.854 |      3.949 |      4.619 |      4.854 |     20.380 |
mean:   -   3.45%
median: -   3.48%
Likely significant

Q: select * from sys.shards
C: 15
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        7.579 ±    2.682 |      4.228 |      6.769 |      8.906 |     24.615 |
|   V2    |        7.925 ±    2.263 |      4.398 |      7.399 |      8.737 |     21.500 |
mean:   +   4.47%
median: +   8.89%
Likely significant
```

### CODE REVIEW

File: sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public ShardRowContext(BlobShard blobShard, ClusterService clusterService) {
+        this(blobShard.indexShard(), blobShard, clusterService, blobShard::getTotalSize);
+    }
+
+    private ShardRowContext(IndexShard indexShard,
+                            @Nullable BlobShard blobShard,
+                            ClusterService clusterService,
+                            Supplier<Long> sizeSupplier) {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java]: what size is this? also is this used? didn't find any usage of the `size` method

File: sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import org.elasticsearch.common.lucene.BytesRefs;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.store.StoreStats;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class ShardRowContext {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java]: Would this better be named as `ShardContext`? Not sure if there is any row relevant information in there

File: sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java (hunk 1 above)
Comment (seut) [reply sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java]: see https://github.com/crate/crate/pull/7594/files#diff-5a4815a0678a6f35b2fa4aaa3aa8742eR141

File: sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java (hunk 2 above)
Comment (matriv) [reply sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java]: or `SysShardContext` ?

Comment (seut) [reply sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java]: this context is used for a shard **row** not for a shard, so I choose this to make it clear ;)

Comment (seut) [reply sql/src/main/java/io/crate/expression/reference/sys/shard/ShardRowContext.java]: I don't see how just `RowContext` will make it more clear, but I'm totally fine with this. not sure what I should document here more that what the class name is telling... (its a context for shard rows)

**Pull request merged**